File: FPTalk1/Content.cs
using DiscriminatedUnion;
using DiscriminatedUnion.Json;
using Microsoft.AspNetCore.Http;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
namespace FPTalk1
{
	public static class Content
	{
		public static Func<HttpContext, Func<Task>, Task> Middleware(
			Func<HttpContext, bool> filter,
			Func<HttpContext, Task> act) =>
			(cont, continuewith) => filter(cont) ? act(cont) : continuewith();

		/// Object to Json 
		public static Stream Json<T>(Stream ms, T myObj)
		{
			using(var writer = new StreamWriter(ms))
			{ 
				writer.Write(JsonConvert.SerializeObject(myObj));
			}
			return ms;
		}

		///// Object from Json 
		public static T Unjson<T>(Stream jsonStream)
		{
			var reader = new StreamReader(jsonStream);
			return JsonConvert.DeserializeObject<T>(reader.ReadToEnd());
		}
		public static Func<A, B> fun<A, B>(Func<A, B> f) => f;

		// Function Adapter
		public static Func<HttpContext, Task> Jsonify<A, B>(this Func<A, B> f) =>
			(HttpContext a) => Task.Run(() => Json(a.Response.Body, f(Unjson<A>(a.Request.Body))));

		// Routing Filters
		public static Func<HttpContext, bool> PathFilter(PathString path) =>
			(HttpContext h) =>
				h.Request.Path == path;

		public static Func<HttpContext, bool> MethodFilter(RequestMethod method) =>
			(HttpContext h) => 
				 h.Request.Method == Enum.GetName(typeof(RequestMethod), method);

		public static Func<HttpContext, bool> And(this Func<HttpContext, bool> a, Func<HttpContext, bool> b) =>
			(HttpContext h) => a(h) && b(h);

	}
}

File: FPTalk2/Content.fs
module Content

open System.Threading.Tasks
open System
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Builder
open System.Runtime.Serialization.Json
open System.IO
open System.Text
open Newtonsoft.Json

type RequestMethod =
    | Trace
    | Put
    | Post
    | Patch
    | Options
    | Head
    | Get
    | Delete
    | Connect

type Skip = Func<Task>

type Filter = 
    HttpContext-> bool

type RequestHandler = 
    HttpContext -> Task

type Middleware = 
    Filter -> RequestHandler -> HttpContext -> Skip -> Task

type Adapter<'a, 'b> = ('a -> 'b) -> (HttpContext -> Task)

/// Object to Json 
let internal json<'t> (ms:Stream) (myObj:'t) =
    use writer = new StreamWriter(ms)
    writer.Write(JsonConvert.SerializeObject(myObj))
    ms

/// Object from Json 
let internal unjson<'t> (jsonStream:Stream)  : 't =
    use reader = new StreamReader(jsonStream)
    JsonConvert.DeserializeObject<'t>(reader.ReadToEnd())

let useEx app f = UseExtensions.Use(app, new Func<HttpContext, Func<Task>, Task>(f)) |> ignore

let Middleware:Middleware = fun filter b context continuewith -> 
    if filter context then b context else continuewith.Invoke()

let Jsonify:Adapter<'a, 'b> = (fun f a -> 
    json a.Response.Body (f (unjson a.Request.Body)) |> ignore
    Task.CompletedTask)

let PathFilter (path:String) (h:HttpContext) =  h.Request.Path = new PathString(path)

let MethodFilter (method:RequestMethod) (h:HttpContext) = //true
    match method with
        | RequestMethod.Trace -> h.Request.Method = "TRACE"
        | RequestMethod.Put -> h.Request.Method = "PUT"
        | RequestMethod.Post -> h.Request.Method = "POST"
        | RequestMethod.Patch -> h.Request.Method = "PATCH"
        | RequestMethod.Options -> h.Request.Method = "OPTIONS"
        | RequestMethod.Head -> h.Request.Method = "HEAD"
        | RequestMethod.Get -> h.Request.Method = "GET"
        | RequestMethod.Delete -> h.Request.Method = "DELETE"
        | RequestMethod.Connect -> h.Request.Method = "CONNECT"

//method.ToString()

let ContentFilter (content:string) (h:HttpContext) = 
    h.Request.ContentType = content

